### ISSUE TITLE
[JavaScript] Implement new Concept Exercise: object-destructuring

### ISSUE THREAD

**Issue by SleeplessByte** [exercism/v3#1894]
This issue describes how to implement the `object-destructuring` concept exercise for the `javascript` track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

See the documentation _above_ (general documentation), as well as [How to implement a Concept Exercise in JavaScript](https://github.com/exercism/v3/blob/master/languages/javascript/reference/implementing-a-concept-exercise.md).

## Goal

The goal of this exercise is to teach the student how they can _destructure_ an object into multiple bindings (which looks like pattern matching).

## Learning objectives

- Using destructuring to get a property
- Using destructuring to get a nested property
- Using destructurng to get multiple properties ([object pattern shorthand](https://exploringjs.com/impatient-js/ch_destructuring.html#property-value-shorthands))
- Using destructuring to _rename_ properties
- Using destructuring + [rest properties](https://exploringjs.com/impatient-js/ch_destructuring.html#rest-properties) to extract a property, and put the rest of the properties in a new name (after.md: React has the pattern  `...otherProps` to pass on a few props, but not all).
- Using [default values in object destructuring](https://exploringjs.com/impatient-js/ch_destructuring.html#default-values-in-object-destructuring)
- Using [nested destructuring](https://exploringjs.com/impatient-js/ch_destructuring.html#nested-destructuring).

## Out of scope

- Anything _specific_ with arrays because they already know this

## Concepts

- `object-destructuring`
- `rest-properties`

@ErikSchierboom please advice. These are two concepts, which can be *used* on their own, but not when learning/teaching. It makes absolute sense to have them _separately_ for exercises, but not when teaching the concept itself.

However, we _could_ argue that, because you learn them all together here, that for us it's just 'one concept'. In that case, I need help with the name (probably still `object-descructuring`). The counter part is `array-destructuring`).

## Prerequisites

- `objects`
- `array-destructuring`
- ... (do you think something is missing?)

### After

- Usage in the parameters directly,
- Usage when extracting data from e.g. a JSON response
- Usage in e.g. React

## Help

You can choose to do this solo-style, or collaborate with multiple people on this. My suggestion would be is to 

1. **first accept this issue by saying "I'd like to work on this"** (you will not get a response from me. Just go with it) and optionally request that someone works with you (and wait for a second person to accept your request).
2. Use this issue to write down what you need (collect reference material) and discuss, if necessary, what to write and what not to write.
3. Draft a PR and request the feedback from at least one other JavaScript member (I recommend just pinging the channel on slack).

**Comment by rishiosaur** [comment]
I'd like to work on this.

**Comment by rishiosaur** [comment]
TODO:

- [ ] Using destructuring to get a property

- [ ] Using destructuring to get a nested property

- [ ] Using destructurng to get multiple properties (object pattern shorthand)

- [ ] Using destructuring to rename properties

- [ ] Using destructuring + rest properties to extract a property, and put the rest of the properties in a new name (after.md: React has the pattern ...otherProps to pass on a few props, but not all).

- [ ] Using default values in object destructuring

- [ ] Using nested destructuring.

**Comment by ErikSchierboom** [comment]
> @ErikSchierboom please advice. These are two concepts, which can be used on their own, but not when learning/teaching. It makes absolute sense to have them separately for exercises, but not when teaching the concept itself.
> 
> However, we could argue that, because you learn them all together here, that for us it's just 'one concept'. In that case, I need help with the name (probably still object-descructuring). The counter part is array-destructuring).

@SleeplessByte I'd have the exercise be named after the _primary_ concept (which looks to be `object-destructuring`) and have it unlock two concepts: `object-destructuring` and `rest-properties`.

**Comment by SleeplessByte** [comment]
Perfect.

**Comment by SleeplessByte** [comment]
@rishiosaur How's it going with this? Need help?

**Comment by evelynstender** [comment]
@rishiosaur are you still working on this? I'd like to help!
